### PR TITLE
fix(express): listener leak

### PR DESCRIPTION
**Pull request by johanneswuerbach**
## Which problem is this PR solving?

Fixes https://github.com/open-telemetry/opentelemetry-js-contrib/issues/258

## Short description of the changes

When a middleware is calling the callback synchronous inside the middleware the span was already ended and the added `finish` listener below was never removed.

### CODE REVIEW

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (johanneswuerbach) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:110]: I'm not sure about this test, it fails without the fix, but it feels somewhat brittle. 

You can replicate the reported warning by registering the middleware above 11 times, which will print `MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 finish listeners added to [ServerResponse]. Use emitter.setMaxListeners() to increase limit`.

**Review by vmarchaud — APPROVED** [review]

**Review by naseemkullah — APPROVED** [review]

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: Well it just make sure we are only adding a specific number of listener, one for each async middleware. 
Your changes only listen for `finish` for async middleware whereas it was for all middlewares before (my mistake) so if the user have more than 10 async middleware, it will still generate that warning.

Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: In this case wouldn't it be better to just use one listener and keep track of started spans? In case the listeners fires we could iterate through the list and close everything, which isn't closed yet.

Comment (vmarchaud) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: Yeah i agree that it would be much better

Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: After re-reading the code, I don't think we have an issue here.

According to the documentation under https://expressjs.com/en/guide/using-middleware.html:

> If the current middleware function does not end the request-response cycle, it must call next() to pass control to the next middleware function. Otherwise, the request will be left hanging.

As the `finish` listener is removed when calling `next()` and middlewares are called in serial, I don't think we can end up with more than one otel `finish` listener in this case.

Comment (vmarchaud) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: There are different possibilities:
1. sync middleware calling `next`
2. sync middleware ending the response
3. async middleware calling `next`
4. async middleware ending the reponse

With your fix, we correctly handle 1, 2 and 4. However when the middleware is doing async work before calling `next`, `spanHasEnded` is `false` since that part is run synchronously after calling the middleware, therefore its adding a listener.

I think storing spans is the only way to do this properly :/

Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts]: Sorry for the late reply, I now added tests for all cases you described, but can't reproduce the problem in any of them (also not in 3).

Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:135]: @vmarchaud could you take another look if I missed anything in those test cases? Thank you 🙇

Comment (vmarchaud) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:135]: @johanneswuerbach Looks good to me

**Review by vmarchaud — APPROVED** [review]

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (obecny) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:266]: please create an arrow function out of the loop scope and then just use it here, instead of creating 15 new callbacks functions

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (obecny) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:227]: the same pattern here

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (obecny) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:187]: and here

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (obecny) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:122]: ditto

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts
Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:266]: I’m not entirely sure how the refactored code should look like. Could you elaborate? The test intentionally creates 15 middlewares to validate we don’t see any warnings with > 11 listeners.

While the middleware function could be shared, it somehow feels less real world like as usually each middleware is a different function with different functionality.

Comment (obecny) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:266]: you declare a function out of the loop and then use it inside loop.

Comment (johanneswuerbach) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:266]: I get that, but then we are calling effectively the same middleware multiple times, which isn’t really like something somebody would do in a real application.

Comment (dyladan) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:266]: I agree with johannes here. I don't think we need to worry too much about micro-optimizing the test code.

**Review by dyladan — APPROVED** [review]

**Comment by johanneswuerbach** [comment]
Fixed the merge conflicts, anything else required before getting this merged?

**Comment by dyladan** [comment]
@johanneswuerbach I see there are open comments from @obecny. If they are resolved please mark them as resolved. If not, please make the suggested fixes or explain your reasoning not to. We typically do not merge any PRs that have open conversations so we can ensure all concerns are addressed before merging.

**Comment by johanneswuerbach** [comment]
@dyladan all four comments are about the extracting the fn used in the loop or keeping it inside, which is being discussed here https://github.com/open-telemetry/opentelemetry-js-contrib/pull/278#discussion_r552698032.

As an outside contributor I'm not really sure whether your agreement to my argument and the one +1 is resolving the issue (and the other 3) or not and as it feels to me like a decision all involved maintainers need to agree upon. As already said I'm happy yo change the code, it just feels wrong, but I'm also not the person who is maintaining the code base so it isn't my call to make :-).

**Comment by obecny** [comment]
> @dyladan all four comments are about the extracting the fn used in the loop or keeping it inside, which is being discussed here [#278 (comment)](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/278#discussion_r552698032).
> 
> As an outside contributor I'm not really sure whether your agreement to my argument and the one +1 is resolving the issue (and the other 3) or not and as it feels to me like a decision all involved maintainers need to agree upon. As already said I'm happy yo change the code, it just feels wrong, but I'm also not the person who is maintaining the code base so it isn't my call to make :-).

eslint has a rule `no-loop-func` which we don't have yet unfortunately (we should imho), once we enable it this code will need to be changed - someone will have to do it. Declaring a new function in a loop is a bad practice as it creates a new function with each iteration. So this isn't even about "micro-optimizing " this is just basic principle of "good code" which is even supported by the mentioned eslint rule - I'm surprised we don't have it enable by default as it should be recommended default and we would not have this discussion as the build would fail then. Having said I think that fixing this issue should be a 1 minute job as it is pretty straight forward and it will make sure the code quality is fine - even in test :).

**Comment by johanneswuerbach** [comment]
👍  updated.

**Review by obecny — APPROVED** [review]
lgtm, thx for changes

**Pull request merged**
